### PR TITLE
Refactor pg_type out of DatabaseCatalog.

### PR DESCRIPTION
**Pull request by lmwnshn**
Refactor `pg_type` logic away from `DatabaseCatalog` into its own `PgTypeImpl` object.

Same deal as #1347 which should be merged first.

Just opening it to keep track of merge order.

### CODE REVIEW

File: src/include/catalog/postgres/pg_type.h [hunk 1]
```diff
@@ -4,39 +4,56 @@
...
+  static constexpr col_oid_t TYPBYVAL_COL_OID = col_oid_t(5);      // BOOLEAN
+  static constexpr col_oid_t TYPTYPE_COL_OID = col_oid_t(6);       // CHAR
+
+  static constexpr uint8_t NUM_PG_TYPE_COLS = 6;
+
+  static constexpr std::array<col_oid_t, NUM_PG_TYPE_COLS> PG_TYPE_ALL_COL_OIDS = {
+      TYPOID_COL_OID, TYPNAME_COL_OID, TYPNAMESPACE_COL_OID, TYPLEN_COL_OID, TYPBYVAL_COL_OID, TYPTYPE_COL_OID};
+
+  // TODO(tanujnay112): Not sure how to do this without hardcoding
+  static constexpr type_oid_t VAR_ARRAY_OID = type_oid_t(11);
```
Comment (jrolli) [inline src/include/catalog/postgres/pg_type.h:56]: I think I remember concluding that the hard-coding was appropriate so I think we can remove the TODO.  However, we probably need to dig into where `11` comes from and document it in a comment here...

File: src/include/catalog/postgres/pg_type_impl.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  /**
+   * Prepare to create pg_type.
+   * Does NOT create anything until the relevant bootstrap functions are called.
+   *
+   * @param db_oid          The OID of the database that pg_type should be created in.
+   */
+  explicit PgTypeImpl(db_oid_t db_oid);
+
+  /** Bootstrap the projected row initializers for pg_type. */
+  void BootstrapPRIs();
```
Comment (jrolli) [inline src/include/catalog/postgres/pg_type_impl.h:38]: Make private and call from `Bootstrap`

File: src/include/catalog/postgres/pg_type_impl.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+  /**
+   * Bootstrap:
+   *    pg_type
+   *    pg_type_oid_index
+   *    pg_type_name_index
+   *    pg_type_namespace_index
```
Comment (jrolli) [inline src/include/catalog/postgres/pg_type_impl.h:45]: Add a description of dependency relations with other tables.

**Review by jrolli — CHANGES_REQUESTED** [review]
Same high-level updates needed as #1344, #1345, and #1347.

File: src/include/catalog/postgres/pg_type_impl.h (hunk 3 above)
Comment (lmwnshn) [reply src/include/catalog/postgres/pg_type_impl.h:45]: wip

File: src/include/catalog/postgres/pg_type_impl.h (hunk 2 above)
Comment (lmwnshn) [reply src/include/catalog/postgres/pg_type_impl.h:38]: unable to at present

Comment (lmwnshn) [reply src/include/catalog/postgres/pg_type_impl.h:38]: Per previous conversation, might not be doable until builder is refactored

File: src/include/catalog/postgres/pg_type.h (hunk 1 above)
Comment (lmwnshn) [reply src/include/catalog/postgres/pg_type.h]: Please see [this](https://github.com/cmu-db/noisepage/pull/1348/commits/34bfece857ca44c46377c98192b4ac1d322af00b), thoughts?

**Comment by lmwnshn** [comment]
Will merge with #1358.

**Pull request closed without merging**